### PR TITLE
First proper pass at link canonicalization

### PR DESCRIPTION
**Pull request by jskeet**
This is ludicrously complicated, but that's because the packages in
googleapis.dev just try to include everything that's needed... and
because some of our packages (e.g. Google.Apis and Google.Apis.Core)
aren't brilliant in terms of consistency.

There may still be corner cases that I'm missing. Ideally we'll run
a long-running test that checks the validity of these before doing
anything irreversible.

### CODE REVIEW

File: tools/Google.Cloud.Tools.GenerateCanonicalLinks.Tests/CanonicalizerTest.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        [InlineData("Google.Apis", "api/Google.Apis.ETagAction.html", "Google.Apis/latest/Google.Apis.ETagAction")]
+        [InlineData("Google.Apis", "api/Google.Apis.Download.DownloadStatus.html", "Google.Apis/latest/Google.Apis.Download.DownloadStatus")]
+        [InlineData("Google.Cloud.Asset.V1", "api/Google.Cloud.Iam.V1.AuditConfigDelta.html", "Google.Cloud.Iam.V1/latest/Google.Cloud.Iam.V1.AuditConfigDelta")]
+        [InlineData("Google.Cloud.Logging.V2", "api/Google.Cloud.Logging.Type.HttpRequest.html", "Google.Cloud.Logging.Type/latest/Google.Cloud.Logging.Type.HttpRequest")]
+        [InlineData("Google.Cloud.Firestore", "api/Google.Cloud.Firestore.V1.Firestore.html", "Google.Cloud.Firestore.V1/latest/Google.Cloud.Firestore.V1.Firestore")]
+        [InlineData("Google.Cloud.Asset.V1", "api/Google.Cloud.OrgPolicy.V1.Policy.html", "Google.Cloud.OrgPolicy.V1/latest/Google.Cloud.OrgPolicy.V1.Policy")]
+        [InlineData("Google.Cloud.Asset.V1", "api/Google.Identity.AccessContextManager.Type.DeviceEncryptionStatus.html", "Google.Identity.AccessContextManager.Type/latest/Google.Identity.AccessContextManager.Type.DeviceEncryptionStatus")]
```
Comment (amanda-tarafa) [inline tools/Google.Cloud.Tools.GenerateCanonicalLinks.Tests/CanonicalizerTest.cs:41]: Would it make sense to try with something Diagnostics? I'm thinking of two cases (which might already work):
- Something from Diagnostics.Common being used in Diagnostics.AspNetCore.
- And how we have one solution only for Diagnostics.AspNetCore[3] with pragma if for setting one namespace or the other. So maybe try the same type in both Diagnostics.AspNetCore and Diagnostics.AspNetCore3?

**Review by amanda-tarafa — COMMENTED** [review]
LGTM. Just a suggestion for more tests cases maybe.

**Review by amanda-tarafa — APPROVED** [review]

File: tools/Google.Cloud.Tools.GenerateCanonicalLinks.Tests/CanonicalizerTest.cs (hunk 1 above)
Comment (jskeet) [reply tools/Google.Cloud.Tools.GenerateCanonicalLinks.Tests/CanonicalizerTest.cs:41]: Ooh, I'd definitely missed Diagnostics.Common, yes. Okay, will add tests for those and then fix them. Thanks!

Comment (jskeet) [reply tools/Google.Cloud.Tools.GenerateCanonicalLinks.Tests/CanonicalizerTest.cs:41]: Added Common tests and fixed them... AspNetCore and AspNetCore3 should be okay, but the published docs are currently broken - I need to investigate that separately.
